File: backend/backup/feature-flag-service/python/feature_flag_service/model/evaluation_context.py
"""
Evaluation context module for the Feature Flag Service.
"""

from typing import Dict, Optional

from pydantic import BaseModel, Field


class ServiceContext(BaseModel):
    """
    Model for a service context.
    """
    service_name: Optional[str] = None
    service_version: Optional[str] = None
    instance_id: Optional[str] = None
    environment: Optional[str] = None
    metrics: Dict[str, str] = Field(default_factory=dict)

    def set_metric(self, key: str, value: str) -> None:
        """
        Set a service metric.
        """
        self.metrics[key] = value


class EvaluationContext(BaseModel):
    """
    Model for an evaluation context.
    """
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    attributes: Dict[str, str] = Field(default_factory=dict)
    numeric_attributes: Dict[str, float] = Field(default_factory=dict)
    boolean_attributes: Dict[str, bool] = Field(default_factory=dict)
    service_context: Optional[ServiceContext] = None

    def to_map(self) -> Dict:
        """
        Convert the evaluation context to a map for rule evaluation.
        """
        result = {}

        # Add user and session IDs
        if self.user_id:
            result["userId"] = self.user_id
        if self.session_id:
            result["sessionId"] = self.session_id

        # Add attributes
        result.update(self.attributes)
        result.update(self.numeric_attributes)
        result.update(self.boolean_attributes)

        # Add service context
        if self.service_context:
            service = {}
            
            if self.service_context.service_name:
                service["name"] = self.service_context.service_name
            if self.service_context.service_version:
                service["version"] = self.service_context.service_version
            if self.service_context.instance_id:
                service["instanceId"] = self.service_context.instance_id
            if self.service_context.environment:
                service["environment"] = self.service_context.environment
            
            # Add metrics
            if self.service_context.metrics:
                service["metrics"] = self.service_context.metrics
            
            result["service"] = service

        return result

    def set_attribute(self, key: str, value: str) -> None:
        """
        Set a string attribute.
        """
        self.attributes[key] = value

    def set_numeric_attribute(self, key: str, value: float) -> None:
        """
        Set a numeric attribute.
        """
        self.numeric_attributes[key] = value

    def set_boolean_attribute(self, key: str, value: bool) -> None:
        """
        Set a boolean attribute.
        """
        self.boolean_attributes[key] = value

    def set_service_context(self, service_context: ServiceContext) -> None:
        """
        Set the service context.
        """
        self.service_context = service_context

File: backend/backup/feature-flag-service/go/internal/model/evaluation_context.go
package model

// EvaluationContext represents the context for evaluating a feature flag
type EvaluationContext struct {
	UserID            string                 `json:"userId,omitempty"`
	SessionID         string                 `json:"sessionId,omitempty"`
	Attributes        map[string]string      `json:"attributes,omitempty"`
	NumericAttributes map[string]float64     `json:"numericAttributes,omitempty"`
	BooleanAttributes map[string]bool        `json:"booleanAttributes,omitempty"`
	ServiceContext    *ServiceContext        `json:"serviceContext,omitempty"`
}

// ServiceContext represents the service context for evaluation
type ServiceContext struct {
	ServiceName    string            `json:"serviceName,omitempty"`
	ServiceVersion string            `json:"serviceVersion,omitempty"`
	InstanceID     string            `json:"instanceId,omitempty"`
	Environment    string            `json:"environment,omitempty"`
	Metrics        map[string]string `json:"metrics,omitempty"`
}

// NewEvaluationContext creates a new evaluation context
func NewEvaluationContext() *EvaluationContext {
	return &EvaluationContext{
		Attributes:        make(map[string]string),
		NumericAttributes: make(map[string]float64),
		BooleanAttributes: make(map[string]bool),
	}
}

// NewServiceContext creates a new service context
func NewServiceContext() *ServiceContext {
	return &ServiceContext{
		Metrics: make(map[string]string),
	}
}

// ToMap converts the evaluation context to a map for rule evaluation
func (c *EvaluationContext) ToMap() map[string]interface{} {
	result := make(map[string]interface{})

	// Add user and session IDs
	if c.UserID != "" {
		result["userId"] = c.UserID
	}
	if c.SessionID != "" {
		result["sessionId"] = c.SessionID
	}

	// Add attributes
	for k, v := range c.Attributes {
		result[k] = v
	}

	// Add numeric attributes
	for k, v := range c.NumericAttributes {
		result[k] = v
	}

	// Add boolean attributes
	for k, v := range c.BooleanAttributes {
		result[k] = v
	}

	// Add service context
	if c.ServiceContext != nil {
		service := make(map[string]interface{})
		
		if c.ServiceContext.ServiceName != "" {
			service["name"] = c.ServiceContext.ServiceName
		}
		if c.ServiceContext.ServiceVersion != "" {
			service["version"] = c.ServiceContext.ServiceVersion
		}
		if c.ServiceContext.InstanceID != "" {
			service["instanceId"] = c.ServiceContext.InstanceID
		}
		if c.ServiceContext.Environment != "" {
			service["environment"] = c.ServiceContext.Environment
		}
		
		// Add metrics
		metrics := make(map[string]interface{})
		for k, v := range c.ServiceContext.Metrics {
			metrics[k] = v
		}
		
		if len(metrics) > 0 {
			service["metrics"] = metrics
		}
		
		result["service"] = service
	}

	return result
}

// SetAttribute sets a string attribute
func (c *EvaluationContext) SetAttribute(key, value string) {
	c.Attributes[key] = value
}

// SetNumericAttribute sets a numeric attribute
func (c *EvaluationContext) SetNumericAttribute(key string, value float64) {
	c.NumericAttributes[key] = value
}

// SetBooleanAttribute sets a boolean attribute
func (c *EvaluationContext) SetBooleanAttribute(key string, value bool) {
	c.BooleanAttributes[key] = value
}

// SetServiceContext sets the service context
func (c *EvaluationContext) SetServiceContext(serviceContext *ServiceContext) {
	c.ServiceContext = serviceContext
}

// SetServiceMetric sets a service metric
func (s *ServiceContext) SetMetric(key, value string) {
	s.Metrics[key] = value
}

